File: Mootata/34회차/25644.최대 상승.py
n = int(input())
values = list(map(int, input().split()))
min_val = values[0]
answer = 0

for v in values:
    if min_val < v:
        answer = max(answer, v - min_val)
    else:
        min_val = v

print(answer)

# 날짜별 주가를 확인하여 min_val보다 값이 크면, 이득을 볼 수 있다는 뜻이므로 answer 초기화
# min_val보다 값이 작다면 해당 가격으로 구매하는 것이 더 큰 이득을 볼 수 있으므로 min_val 초기화

File: apua34/2회차/25644.최대상승.js
/**
 * @file 25644.최대상승.js
 * @author @apua34
 * @brief 최대 상승
 * @link https://www.acmicpc.net/problem/25644
 * @version 0.1
 * @date 2023-07-19
 * @strategy 그리디
 *
 * @copyright Copyright (c) 2023 @apua34
 *
 */

const fs = require("fs");
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
const input = fs.readFileSync(filePath)
    .toString()
    .trim()
    .split("\n")

const [N, ...arr] = input;
let numberArr = arr.map((item) => item.split(" ").map(Number))[0];

function solution(N, arr) {
    let answer = 0;
    let elementStock = 0;

    // ex) N = 5, 뒤에서 부터 진행
    for(let i = N - 1; i >= 0; i--) { 
        elementStock = Math.max(elementStock, arr[i]); // 0, 5
        answer = Math.max(answer, elementStock - arr[i]);  // 주가 공식 : 0 - 5
    }
    
    return answer;
}

const check = solution(N, numberArr);
console.log(check);


